File: 1289-minimum-falling-path-sum-ii/1289-minimum-falling-path-sum-ii.java
class Solution {
    public int minFallingPathSum(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        List<List<Integer>> dp = new ArrayList<>(m);
        for (int i = 0; i < m; i++) {
            dp.add(new ArrayList<>(Collections.nCopies(n, 0)));
        }
        for (int i = 0; i < n; i++) {
            dp.get(0).set(i, grid[0][i]);
        }
        for (int r = 1; r < m; r++) {
            for (int c = 0; c < n; c++) {
                int min_value = Integer.MAX_VALUE;
                for (int k = 0; k < n; k++) {
                    if (k == c) {
                        continue;
                    }
                    min_value = Math.min(min_value, dp.get(r - 1).get(k));
                }
                dp.get(r).set(c, grid[r][c] + min_value);
            }
        }
        return Collections.min(dp.get(m - 1));
    }
}

File: 1289-minimum-falling-path-sum-ii/1289-minimum-falling-path-sum-ii.cpp
class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size();
        vector<vector<int>> dp(m, vector<int>(n, 0));
        
        for(int i = 0; i < n; ++i) dp[0][i] = grid[0][i];
        
        for(int r = 1; r < m; ++r)
            for(int c = 0; c < n; ++c) {
                int min_value = INT_MAX;
                for(int k = 0; k < n; ++k) {
                    if(k == c) continue;
                    min_value = min(min_value, dp[r - 1][k]);
                }
                dp[r][c] = grid[r][c] + min_value;
            }
        
        return *min_element(dp[m - 1].begin(), dp[m - 1].end());
    }
};